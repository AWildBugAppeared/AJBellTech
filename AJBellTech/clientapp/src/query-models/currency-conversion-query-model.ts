export interface CurrencyConversionQueryModel {
  amount: number;
  currency: string;
}