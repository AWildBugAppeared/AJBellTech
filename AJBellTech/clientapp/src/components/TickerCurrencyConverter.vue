<template>
  <div>
    <v-form id="currencyConversionForm" ref="form" @submit.prevent="submit">
      <v-row justify="center">
        <v-col cols="3">
          <v-select
            v-model="selectedCurrency"
            :items="validCurrenciesForBtcConversion"
            :rules="[requiredRule]"
            label="Select Currency"
          ></v-select>
        </v-col>
        <v-col cols="3">
          <v-text-field
            v-model="amount"
            label="Enter Amount"
            type="number"
            :rules="[requiredRule]"
          ></v-text-field>
        </v-col>
        <v-col cols="2" align-self="center">
          <v-btn
            :loading="isLoading"
            form="currencyConversionForm"
            type="submit"
            @click="submit"
            >Convert to Btc</v-btn
          >
        </v-col>
      </v-row>
    </v-form>
    <v-row v-if="errorMessage" justify="center" data-error>
      <v-col cols="8" class="red--text">{{ errorMessage }}</v-col>
    </v-row>
    <v-row
      v-if="btcAmount"
      justify="center"
      class="green--text"
      data-conversion-result
    >
      <v-col cols="8">
        {{ currencyAmountConverted }} units of {{ selectedCurrency }} equals
        {{ btcAmount }} in BTC
      </v-col>
    </v-row>
  </div>
</template>

<script lang="ts">
import { mapActions, mapState } from "vuex";
import Vue from "vue";
import { CurrencyConversionQueryModel } from "@/query-models/currency-conversion-query-model";

export default Vue.extend({
  name: "TickerCurrencyConverter",

  data() {
    return {
      amount: 0,
      btcAmount: 0,
      currencyAmountConverted: 0,
      errorMessage: "",
      isLoading: false,
      requiredRule: (v: string | number) => !!v || "This field is required",
      selectedCurrency: "",
    };
  },

  computed: {
    ...mapState("ticker", ["validCurrenciesForBtcConversion"]),
  },

  methods: {
    ...mapActions("ticker", ["getBtcAmountFromCurrency"]),

    submit() {
      const isValid = (this.$refs.form as Vue & {
        validate: () => boolean;
      }).validate();

      if (!isValid) return;

      this.isLoading = true;
      this.btcAmount = 0;
      this.currencyAmountConverted = 0;
      this.errorMessage = "";

      var payload: CurrencyConversionQueryModel = {
        amount: this.amount,
        currency: this.selectedCurrency,
      };

      this.getBtcAmountFromCurrency(payload)
        .then((amount: number) => {
          this.currencyAmountConverted = this.amount;
          this.btcAmount = amount;
        })
        .catch((res) => {
          this.errorMessage = res.response.data;
        })
        .finally(() => {
          this.isLoading = false;
        });
    },
  },
});
</script>

<style lang="scss" scoped></style>
