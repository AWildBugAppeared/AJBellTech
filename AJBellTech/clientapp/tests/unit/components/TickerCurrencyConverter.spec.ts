/* eslint-disable  @typescript-eslint/no-explicit-any */

import Vue from "vue";
import Vuetify from "vuetify";
import Vuex from "vuex";
import { createLocalVue, mount, Wrapper } from "@vue/test-utils";

import TickerCurrencyConverter from "@/components/TickerCurrencyConverter.vue";
import { CurrencyConversionQueryModel } from "@/query-models/currency-conversion-query-model";

Vue.use(Vuex);

describe("TickerCurrencyConverter.vue", () => {
  let actions: {
    getBtcAmountFromCurrency: jest.Mock;
  };

  const btcAmountReturnedFromApi = 0.001;

  const localVue = createLocalVue();
  let wrapper: Wrapper<Vue & { [key: string]: any }>;
  let mountWrapper: () => void;

  beforeEach(() => {
    const vuetify = new Vuetify({
      mocks: {
        $vuetify: {},
      },
    });
    Vue.use(Vuetify);
    document.body.setAttribute("data-app", "true");

    actions = {
      getBtcAmountFromCurrency: jest
        .fn()
        .mockResolvedValue(btcAmountReturnedFromApi),
    };

    mountWrapper = () => {
      const storeData = {
        modules: {
          ticker: {
            state: {
              validCurrenciesForBtcConversion: [
                "GBP",
                "USD",
                "EUR",
                "CAD",
                "AUD",
              ],
            },
            namespaced: true,
            actions: actions as any,
          },
        },
      };
      const $store = new Vuex.Store(storeData);

      wrapper = mount(TickerCurrencyConverter, {
        localVue,
        vuetify,
        data() {
          return {};
        },
        mocks: {
          $store,
        },
      });
    };
  });

  afterEach(() => {
    wrapper.destroy();
    jest.resetModules();
    jest.clearAllMocks();
  });

  describe("rendering", () => {
    it("should render default view", async () => {
      mountWrapper();
      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it("should render error message", async () => {
      mountWrapper();
      wrapper.vm.errorMessage = "error";
      await wrapper.vm.$nextTick();

      const message = wrapper.find("[data-error]");

      expect(message.html()).toMatchInlineSnapshot(`
        <div data-error="" class="row justify-center">
          <div class="red--text col col-8">error</div>
        </div>
      `);
    });

    it("should render conversion message", async () => {
      mountWrapper();
      wrapper.vm.btcAmount = 0.1234;
      wrapper.vm.currencyAmountConverted = 20;
      wrapper.vm.selectedCurrency = "GBP";
      await wrapper.vm.$nextTick();

      const message = wrapper.find("[data-conversion-result]");

      expect(message.html()).toMatchInlineSnapshot(`
        <div data-conversion-result="" class="row green--text justify-center">
          <div class="col col-8">
            20 units of GBP equals
            0.1234 in BTC
          </div>
        </div>
      `);
    });
  });

  describe("methods", () => {
    describe("submit", () => {
      beforeEach(() => {
        mountWrapper();
        (wrapper.vm.$refs.form as any).validate = jest
          .fn()
          .mockReturnValue(true);
      });

      it("should not make api call when form is not valid ", async () => {
        mountWrapper();
        wrapper.vm.submit();

        await wrapper.vm.$nextTick();

        expect(actions.getBtcAmountFromCurrency.mock.calls.length).toBe(0);
      });

      it("should set state to loading when submitting, and reset when finished", async () => {
        wrapper.vm.submit();

        expect(wrapper.vm.isLoading).toBe(true);

        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();

        expect(wrapper.vm.isLoading).toBe(false);
      });

      it("should clear data before making api call", () => {
        wrapper.vm.btcAmount = 10;
        wrapper.vm.currencyAmountConverted = 0;
        wrapper.vm.errorMessage = "error";
        wrapper.vm.submit();

        expect(wrapper.vm.btcAmount).toBe(0);
        expect(wrapper.vm.currencyAmountConverted).toBe(0);
        expect(wrapper.vm.errorMessage).toBe("");
      });

      it("should process successful response", async () => {
        const amount = 20;
        const currency = "GBP";
        const payload: CurrencyConversionQueryModel = {
          amount: amount,
          currency: currency,
        };

        wrapper.vm.amount = amount;
        wrapper.vm.selectedCurrency = currency;

        wrapper.vm.submit();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();

        expect(actions.getBtcAmountFromCurrency.mock.calls.length).toBe(1);
        expect(actions.getBtcAmountFromCurrency.mock.calls[0][1]).toEqual(
          payload
        );
        expect(wrapper.vm.btcAmount).toBe(btcAmountReturnedFromApi);
      });

      it("should process unsuccessful response", async () => {
        let errorMessage = "error";
        actions.getBtcAmountFromCurrency = jest.fn().mockRejectedValue({
          response: {
            data: errorMessage,
          },
        });

        mountWrapper();
        (wrapper.vm.$refs.form as any).validate = jest
          .fn()
          .mockReturnValue(true);

        const amount = 20;
        const currency = "GBP";
        const payload: CurrencyConversionQueryModel = {
          amount: amount,
          currency: currency,
        };

        wrapper.vm.amount = amount;
        wrapper.vm.selectedCurrency = currency;

        wrapper.vm.submit();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();

        expect(wrapper.vm.errorMessage).toBe(errorMessage);
      });
    });
  });
});
