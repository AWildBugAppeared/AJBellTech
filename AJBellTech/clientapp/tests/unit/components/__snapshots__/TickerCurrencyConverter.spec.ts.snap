// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickerCurrencyConverter.vue rendering should render default view 1`] = `
<div>
  <form novalidate="novalidate" id="currencyConversionForm" class="v-form">
    <div class="row justify-center">
      <div class="col col-3">
        <div class="v-input theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-5" class="v-input__slot">
              <div class="v-select__slot"><label for="input-5" class="v-label theme--light" style="left: 0px; position: absolute;">Select Currency</label>
                <div class="v-select__selections"><input id="input-5" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col col-3">
        <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-10" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Enter Amount</label><input id="input-10" type="number"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col col-2 align-self-center"><button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default" form="currencyConversionForm"><span class="v-btn__content">Convert to Btc</span></button></div>
    </div>
  </form>
  <!---->
  <!---->
</div>
`;
